server/service/src/invoice_line/stock_out_line/set_prescribed_quantity/mod.rs

```diff
@@ -0,0 +1,338 @@
+
+mod generate;
+use generate::generate;
+mod validate;
+use validate::validate;
+
+#[derive(Clone, Debug, PartialEq, Default)]
+pub struct SetPrescribedQuantity {
+    pub invoice_id: String,
+    pub item_id: String,
+    pub prescribed_quantity: f64,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum SetPrescribedQuantityError {
+    ItemNotFound, // check if valid item id
+    InvoiceDoesNotExist,
+    NotAStockItem,
+    NotAPrescription,
+    NotThisStoreInvoice,
+    NewlyCreatedLineDoesNotExist,
+    DatabaseError(RepositoryError),
+}
+
+impl From<RepositoryError> for SetPrescribedQuantityError {
+    fn from(error: RepositoryError) -> Self {
+        SetPrescribedQuantityError::DatabaseError(error)
+    }
+}
+
+pub fn set_prescribed_quantity(
+    ctx: &ServiceContext,
+    input: SetPrescribedQuantity,
+) -> Result<InvoiceLine, SetPrescribedQuantityError> {
+    let invoice_line = ctx
+        .connection
+        .transaction_sync(|connection| {
+            let item_row = validate(connection, &ctx.store_id, &input)?;
+
+            let invoice_line_row_repo = InvoiceLineRowRepository::new(connection);
+
+            let existing_lines = InvoiceLineRepository::new(connection).query_by_filter(
+                InvoiceLineFilter::new()
+                    .item_id(EqualFilter::equal_to(&input.item_id))
+                    .invoice_id(EqualFilter::equal_to(&input.invoice_id)),
+            )?;
+
+            if existing_lines.len() > 1 {
+                let has_prescribed_quantity_or_stock_line = existing_lines.iter().any(|line| {
+                    line.invoice_line_row.prescribed_quantity.is_some()
+                        || line.invoice_line_row.stock_line_id.is_some()
+                });
+
+                if has_prescribed_quantity_or_stock_line {
+                    if let Some(unallocated_line) = existing_lines.iter().find(|line| {
+                        line.invoice_line_row.r#type == InvoiceLineType::UnallocatedStock
+                    }) {
+                        invoice_line_row_repo.delete(&unallocated_line.invoice_line_row.id)?;
+                    }
+                }
+            }
+
+            let existing_line_with_stock = existing_lines
+                .iter()
+                .find(|line| line.invoice_line_row.stock_line_id.is_some());
+
+            let new_line;
+            if let Some(existing_line) = existing_line_with_stock {
+                let mut updated_line = existing_line.clone();
+                updated_line.invoice_line_row.prescribed_quantity = Some(input.prescribed_quantity);
+                invoice_line_row_repo.upsert_one(&updated_line.invoice_line_row)?;
+                new_line = updated_line.invoice_line_row
+            } else {
+                let invoice_line = generate(uuid(), item_row, input.clone())?;
+                invoice_line_row_repo.upsert_one(&invoice_line)?;
+                new_line = invoice_line;
+            }
```

Comment:
```suggestion
            let new_line = match existing_line_with_stock {
                Some(existing_line) => {
                    let mut updated_line = existing_line.clone();
                    updated_line.invoice_line_row.prescribed_quantity =
                        Some(input.prescribed_quantity);
                    invoice_line_row_repo.upsert_one(&updated_line.invoice_line_row)?;
                    updated_line.invoice_line_row
                }
                None => {
                    let invoice_line = generate(uuid(), item_row, input.clone())?;
                    invoice_line_row_repo.upsert_one(&invoice_line)?;
                    invoice_line
                }
            };
```

Using a match here is more `Rusty` than the if else.
I'm actually quite surprised your code worked without assigning `new_line` as `mut`?